lambdas/tests/unit/services/test_login_service.py

```diff
@@ -0,0 +1,375 @@
+        500, "Failed to retrieve access token from ID Provider"
+    )
+    login_service = LoginService()
+    mocker.patch.object(
+        LoginService, "have_matching_state_value_in_record", return_value=True
+    )
+
+    with pytest.raises(AuthorisationException):
+        login_service.generate_session("auth_code", "state")
+
+    mock_oidc_service["fetch_user_org_codes"].assert_not_called()
+    mock_aws_infras["session_table"].post.assert_not_called()
+
+
+def test_exchange_token_raises_auth_error_when_given_state_is_not_in_state_table(
+    mock_aws_infras, mock_oidc_service, set_env, context, mocker
+):
+    mocker.patch.object(
+        DynamoDBService, "simple_query", return_value={"Count": 0, "Items": []}
+    )
+
+    login_service = LoginService()
+
+    with pytest.raises(AuthorisationException):
+        login_service.generate_session("auth_code", "state")
+
+    mock_oidc_service["fetch_token"].assert_not_called()
+    mock_oidc_service["fetch_user_org_codes"].assert_not_called()
+    mock_aws_infras["session_table"].post.assert_not_called()
+
+
+def test_exchange_token_raises_auth_error_when_user_doesnt_have_a_valid_role_to_login(
+    mock_aws_infras,
+    mock_oidc_service,
+    mocker,
+):
+    class PermittedOrgs:
+        def keys(self):
+            return []
+
+    dynamo_state_query_result = {"Count": 1, "Items": [{"id": "state"}]}
+
+    mocker.patch.object(
+        DynamoDBService, "simple_query", return_value=dynamo_state_query_result
+    )
+
+    mocker.patch.object(DynamoDBService, "delete_item")
+
+    mocker.patch(
+        "services.ods_api_service.OdsApiService.fetch_organisation_with_permitted_role"
+    ).return_value = PermittedOrgs()
+
+    login_service = LoginService()
+
+    with pytest.raises(AuthorisationException):
+        login_service.generate_session("auth_code", "state")
+
+    mock_aws_infras["session_table"].post.assert_not_called()
+
+
+# TODO needs to go in dynamo service tests when refactoring of CUT is done
+def test_exchange_token_raises_error_when_encounter_boto3_error(
+    mock_aws_infras, set_env, mocker, context
+):
+    mock_aws_infras["state_table"].query.side_effect = ClientError(
+        {"Error": {"Code": "500", "Message": "mocked error"}}, "test"
+    )
+
+    mock_oidc = mocker.patch("services.oidc_service.OidcService.fetch_oidc_parameters")
+    mock_oidc.return_value = {
+        "OIDC_CLIENT_ID": "client-id",
+        "OIDC_CLIENT_SECRET": "client-secret",
+        "OIDC_ISSUER_URL": "https://issuer-url.com",
+        "OIDC_TOKEN_URL": "https://token-url.com",
+        "OIDC_USER_INFO_URL": "https://userinfo-url.com",
+        "OIDC_JWKS_URL": "https://jwks-url.com",
+        "OIDC_CALLBACK_URL": "https://callback-url.com",
+    }
+
+    login_service = LoginService()
+
+    with pytest.raises(Exception):
+        login_service.generate_session("auth_code", "state")
+    # TODO assert 500 in error
```

Comment:
Has this TODO been addressed?

Reply:
Added assert clauses for the error codes in the exceptions